Start bhima from mocha

mbayopanda opened the pull request:
This PR is related to the issue #4617 

Close #4617

mbayopanda committed:
run app with mocha

mbayopanda committed:
Merge branch 'master' of https://github.com/IMA-WorldHealth/bhima into start-bhima-from-mocha-4617

mbayopanda committed:
integration test from mocha without upload folder

jniles commented on sh/integration-tests.sh:
```diff
@@ -13,24 +13,12 @@ set +a
-TIMEOUT=${BUILD_TIMEOUT:-8}
-
 echo "[test] Spawning server process..."
-# build and start the server
```
👍

jniles commented on sh/integration-tests.sh:
```diff
@@ -13,24 +13,12 @@ set +a
 echo "[test] running tests using mocha"
-
 # run the tests
-../node_modules/.bin/mocha --recursive --bail ../test/integration/
```
Why do we need to put `NODE_ENV` here?  I think it should be fine without it.

jniles replied:
Also, something is keeping our server alive and it isn't exiting.  I've ruled out `lib/db` and `@imaworldhealth/coral`, but still can't figure it out.  Let's not deal with it in this PR.  Instead, let's add a TODO and pass the `--exit` flag.  

Then create an issue and we can work on it after this PR gets merged.

mbayopanda replied:
The reason of `NODE_ENV` is without that mocha cannot read the environment file, so the server is still running in production mode

jniles replied:
It should always read the `.env.development` file.  The code is [here](https://github.com/IMA-WorldHealth/bhima/blob/master/sh/test.sh#L8).  At least it does on my machine and on Travis.  Does it not work for you?

jniles commented on .env.development:
```diff
@@ -11,7 +11,7 @@ DB_NAME='bhima_test'
 SESS_SECRET='XopEn BlowFISH'
 
-DEBUG=app,errors,ReportManager
+DEBUG=app
```
We should probably keep `errors` around at minumum.

jniles commented on server/app.js:
```diff
@@ -67,7 +67,7 @@ function configureServer() {
   // create the server
   http.createServer(app)
-    .listen(process.env.PORT, () => {
+    .listen(port, () => {
```
Nice. :+1:

jniles commented on test/integration/setup.js:
```diff
@@ -10,12 +10,14 @@
+
 // base URL for all tests
 const port = process.env.PORT || 8080;
 const baseUrl = `http://localhost:${port}`;
```
Let's remove all references to baseUrl.  Nothing is using it that I can tell.

jniles commented on test/integration/setup.js:
```diff
@@ -32,17 +34,19 @@ before(() => {
 // runs after all tests are completed
-after(() => {
+after(done => {
+  delete require.cache[require.resolve('../../bin/server/app')];
```
I'm not sure I understand why you need to do this.  It also looks really hacky.  Can we remove it?

mbayopanda replied:
You're totally right, that's the reason why i talked about that

mbayopanda committed:
Applying enhancement after the first review

Close #4617

jniles committed:
fix: remove unused code, add server.close()

Adds the .close() method to the agent in the mocha after() block.
Removes the trap function from the bash script to prevent it screwing up
the integration tests.

mbayopanda committed:
Merge pull request #3 from jniles/start-bhima-from-mocha-4617

fix: remove unused code, add server.close()
